src/vm/eventtrace.cpp

```diff
@@ -3722,6 +3722,75 @@ VOID ETW::TypeSystemLog::DeleteTypeHashNoLock(AllLoggedTypes **ppAllLoggedTypes)
 
 //---------------------------------------------------------------------------------------
 //
+// Called from shutdown to give us the opportunity to dump any sampled object allocation
+// information before the process shuts down.
+//
+
+// static
+VOID ETW::TypeSystemLog::FlushObjectAllocationEvents()
+{
+    LIMITED_METHOD_CONTRACT;
+
+    // If logging is not enabled, then we don't need to do any work.
+    if (!(s_fHeapAllocHighEventEnabledNow || s_fHeapAllocHighEventEnabledNow))
+    {
+        return;
+    }
+
+    AllLoggedTypes * pThreadAllLoggedTypes = NULL;
+    Thread * pThread = NULL;
+
+    // Get the thread store lock.
+    ThreadStoreLockHolder tsl;
+
+    // Iterate over each thread and log any un-logged allocations.
+    while ((pThread = ThreadStore::GetThreadList(pThread)) != NULL)
+    {
+        pThreadAllLoggedTypes = pThread->GetAllocationSamplingTable();
+        if (pThreadAllLoggedTypes == NULL)
+        {
+            continue;
+        }
+
+        DWORD dwAllocsSkippedForSample;
+        SIZE_T cbIgnoredSizeForSample;
+
+        // Iterate over each module.
+        AllLoggedTypesHash * pLoggedTypesHash = &pThreadAllLoggedTypes->allLoggedTypesHash;
```

Comment:
What guarantees that no other threads are modifying these lists? Maybe worth a comment.